docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,309 @@
+ingress --> a --> b & c 
+job --> c
+c --> d & e
+b --> f
+```
+
+We want to bring services into the mesh one by one and trim their config, let's start with service `c` (it's a good example because it has both inbound and outbound traffic):
+
+1. We enable sidecar injection on `c`
+2. The traffic flows through like normal
+3. We define MTP with `allow` from service `a`, for now this has no consequences because there is no service `a` in the Mesh.
+4. We bring service `a` into the mesh
+5. We bring service `job` into the mesh
+6. We switch MTP for `c` to have `allow` from `a` and `job`, then `deny` for any other service (`Mesh`)
+7. This has no consequences on the config size because `a`, `job` and `c` are the only services in the mesh
+8. Now let's bring service `d` into the mesh
+9. When we do that we will observe the first consequence of config trimming,
+   service `d` will have only `a` cluster but won't have `c` cluster because we defined that `c` can only be contacted by `a` and `job`
+10. We repeat this until we have all services in the mesh, all new services that join already have a smaller config because we trimmed it along the way
+
+It's critical to remember to only switch to `deny` when **ALL** consumers are **defined in MTP** and are in the Mesh.
+
+There are other concerns that also need addressing:
+- breaking "top-level targetRef selects proxy to configure" rule
+  - I don't think we can do anything about this, it's at the core of this feature working, what we can (and should) do
+    is to have this feature behind a flag and if there is too much negative feedback about this point we can revert it
+- wrong errors on the client side, instead of 403 client will get 404/"no upstream"
+  - in public facing APIs this is quote common to get a 404 to a resource that you don't have permissions to
+  - I imagine we could have a flag to distinguish between a regular 404 and this 404 caused by no upstream resulting from config trimming and return a 403, but
+    I think this is out of scope for this, probably a future improvement
+- wrong RBAC stats on the server side
+  - we would have correct stats on the client side - if we can translate that and surface it to the user I don't think anyone would mind, but
+    I think its is for future improvement
+- assuming that every service owner knows which services consume their API
+  - as a future improvement we will come up with tooling to assist the users with this problem
+    - have metrics / logs and a dashboard when `allow_with_shadow_deny` is used (this is not a silver bullet - it assumes that during "discovery" period all possible communication will happen, if you gather data for a week and there is a job that runs every quarter it probably won't be recorded)
+
+##### mTLS
+
+When mTLS is disabled, the traffic will go through the `passthrough` cluster and reach the destination.
+When either permissive or strict mTLS is enabled the requests will fail with a closed connection like this:
```

Comment:
Could you also add a response when `passthrough` is disabled and mTLS enabled?

Reply:
will add

Reply:
done